sockeye/inference.py

```diff
@@ -514,156 +547,164 @@ def translate_nd(self, source: mx.nd.NDArray, source_length: mx.nd.NDArray, buck
+        :param source_length: Source length. Shape: (1,).
         :param bucket_key: Bucket key.
         :param max_output_length: Cap the output at this maximum length.
         :return List of lists of word ids, list of attentions, array of accumulated length-normalized
-        negative log-probs.
-        """
-
-        # encode source and initialize decoder states for each model
-        model_encoded_source, model_dynamic_source, model_source_length, model_decoder_states = [], [], [], []
-        model_prev_hidden = []
-
-        for model in self.models:
-            # encode input sentence and initialize decoder states
-            # encoded_source: (self.beam_size, bucket_key, rnn_num_hidden)
-            # decoder_states: [(self.beam_size, rnn_num_hidden),...]
-            encoded_source, source_dynamic_init, source_length, prev_hidden, decoder_states = \
-                model.run_encoder(source,
-                                  source_length,
-                                  bucket_key)
-            model_encoded_source.append(encoded_source)
-            model_dynamic_source.append(source_dynamic_init)
-            model_source_length.append(source_length)
-            model_decoder_states.append(decoder_states)
-            model_prev_hidden.append(prev_hidden)
-
-        # prev_target_word_id(s): (beam_size,)
-        prev_target_word_id = mx.nd.zeros((self.beam_size,), ctx=self.context)
-        prev_target_word_id[:] = self.start_id
-
-        accumulated_scores = mx.nd.zeros((self.beam_size,), ctx=self.context)
-        lengths = mx.nd.zeros((self.beam_size,), ctx=self.context)
-        finished = [False] * self.beam_size
-        sequences = [[] for _ in range(self.beam_size)]
-        # one list of source word attention vectors per hypothesis
-        attention_lists = [[] for _ in range(self.beam_size)]
-        prev_hyp_indices = None
+                negative log-probs.
+        """
+        lengths = mx.nd.zeros((self.beam_size, 1), ctx=self.context)
+        finished = mx.nd.zeros((self.beam_size,), dtype='int32', ctx=self.context)
+        # sequences: (beam_size, output_length)
+        sequences = mx.nd.array(np.full((self.beam_size, max_output_length), C.PAD_ID), dtype='int32', ctx=self.context)
+        # attentions: (beam_size, output_length, bucket_key/source_length)
+        attentions = mx.nd.zeros((self.beam_size, max_output_length, bucket_key), ctx=self.context)
+
+        # best_hyp_indices: row indices of smallest scores (ascending).
+        best_hyp_indices = mx.nd.zeros((self.beam_size,), ctx=self.context)
+        # best_word_indices: column indices of smallest scores (ascending).
+        best_word_indices = mx.nd.zeros((self.beam_size,), ctx=self.context, dtype='int32')
+        # scores_accumulated: chosen smallest scores in scores (ascending).
+        scores_accumulated = mx.nd.zeros((self.beam_size, 1), ctx=self.context)
+
+        self.pad_dist += np.inf
```

Comment:
well, its a potentially large matrix so recreating it for every translation is expensive. We can simply create it once when the translator is created.